packages/arb-rpc-node/cmd/arb-fork-node/arb-fork-node.go

```diff
@@ -0,0 +1,260 @@
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"github.com/rs/zerolog/pkgerrors"
+	golog "log"
+	"math/big"
+	"net/http"
+	_ "net/http/pprof"
+	"os"
+	"path/filepath"
+
+	"github.com/offchainlabs/arbitrum/packages/arb-node-core/cmdhelp"
+	"github.com/offchainlabs/arbitrum/packages/arb-rpc-node/dev"
+)
+
+var pprofMux *http.ServeMux
+
+var logger = arblog.Logger.With().Str("component", "arb-fork-node").Logger()
+
+func init() {
+	pprofMux = http.DefaultServeMux
+	http.DefaultServeMux = http.NewServeMux()
+	gethlog.Root().SetHandler(gethlog.LvlFilterHandler(gethlog.LvlInfo, gethlog.StreamHandler(os.Stderr, gethlog.TerminalFormat(true))))
+}
+
+func main() {
+	// Enable line numbers in logging
+	golog.SetFlags(golog.LstdFlags | golog.Lshortfile)
+
+	// Print stack trace when `.Error().Stack().Err(err).` is added to zerolog call
+	zerolog.ErrorStackMarshaler = pkgerrors.MarshalStack
+
+	zerolog.SetGlobalLevel(zerolog.InfoLevel)
+
+	if err := startup(); err != nil {
+		logger.Error().Err(err).Msg("Error running node")
+	}
+}
+
+func startup() error {
+	ctx, cancelFunc, cancelChan := cmdhelp.CreateLaunchContext()
+	defer cancelFunc()
+
+	fs := flag.NewFlagSet("", flag.ContinueOnError)
+
+	enablePProf := fs.Bool("pprof", false, "enable profiling server")
+	walletcount := fs.Int("walletcount", 10, "number of wallets to fund")
+	walletbalance := fs.Int64("walletbalance", 100, "amount of funds in each wallet (Eth)")
+	dbDir := fs.String("dbdir", "", "directory to load dev node on. Use temporary if empty")
+	aggStr := fs.String("aggregator", "", "aggregator to use as the sender from this node")
+	chainId64 := fs.Uint64("chainId", 68799, "chain id of chain")
+	mnemonic := fs.String(
+		"mnemonic",
+		"jar deny prosper gasp flush glass core corn alarm treat leg smart",
+		"mnemonic to generate accounts from",
+	)
+	fs.Bool("prettyprint", true, "pretty log output")
+	gethLogLevel, arbLogLevel := cmdhelp.AddLogFlags(fs)
+
+	err := fs.Parse(os.Args[1:])
+	if err != nil {
+		return errors.Wrap(err, "error parsing arguments")
+	}
+
+	if err := cmdhelp.ParseLogFlags(gethLogLevel, arbLogLevel); err != nil {
+		return err
+	}
+
+	if *enablePProf {
+		go func() {
+			err := http.ListenAndServe("localhost:8081", pprofMux)
+			log.Error().Err(err).Msg("profiling server failed")
+		}()
+	}
+
+	chainId := new(big.Int).SetUint64(*chainId64)
+
+	wallet, accounts, err := internal.InitializeWallet(*mnemonic, *walletcount)
+	if err != nil {
+		return err
+	}
+
+	var agg common.Address
+	if *aggStr != "" {
+		agg = common.HexToAddress(*aggStr)
+	} else {
+		if len(accounts) < 2 {
+			return errors.New("must have at least 2 accounts")
+		}
+		agg = common.NewAddressFromEth(accounts[1].Address)
+
+	}
+	type forkInfo struct {
+		LastMessage int64 `json:"last_block"`
+	}
+	var fork forkInfo
+
+	forkFile := filepath.Join(*dbDir, "fork.json")
+	forkData, err := os.ReadFile(forkFile)
+	if err == nil {
+		if err := json.Unmarshal(forkData, &fork); err != nil {
```

Comment:
Are we sure we want to reorg the chain every time? It seems useful to be able to preserve state across runs. Maybe we could add a flag to preserve the existing state? Or are we thinking people could just start up the dev node on the same db afterwards to continue their chain?

Reply:
Hmm fair. I was generally just modeling after hardhat's mainnet fork mode, but seems nicer to at least have optional persistance. For now I think just adding a `--persist-state` flag which defaults to off makes sense. For simplicity though I think it makes sense to redo all the initialization message every time which just means your balance will just increase since otherwise it could get into weird inconsistent states